File: algo/sorting/quicksort/qsort.py
import random

def qsort_naive(x):
	'''Quicksort -- O(n log n), O(n) space.
	http://en.wikipedia.org/wiki/Quicksort#Simple_version
	'''

	def partition(x):
		p = x[-1]
		left = []
		right = []
		for item in x[:-1]:
			if item <= p:
				left.append(item)
			else:
				right.append(item)
		return left, right, p

	if len(x) < 2:
		return x
	left, right, p = partition(x)
	return qsort_naive(left) + [p] + qsort_naive(right)

def qsort(x, left, right):
	'''Inplay quicksort -- O(n log n), O(log n) space.
	http://en.wikipedia.org/wiki/Quicksort#Complex_version
	'''

	def swap(x, i, j):
		x[i], x[j] = x[j], x[i]

	def partition(x, left, right, p_idx):
		p_val = x[p_idx]
		swap(x, p_idx, right) # pivot out of the way
		s_idx = left
		for i in xrange(left, right):
			if x[i] <= p_val:
				swap(x, i, s_idx) # put lesser value in store
				s_idx += 1 # advance store
		swap(x, s_idx, right)
		return s_idx

	if right > left:
		p_idx = left + (right - left) / 2 # no overflow
		p_new_idx = partition(x, left, right, p_idx)
		qsort(x, left, p_new_idx - 1)
		qsort(x, p_new_idx + 1, right)


n = 100
big = range(n)
random.shuffle(big)

for a in [
		[],
		[1],
		[1,2],
		[11,1,1,2,3,3,32,2,2,3,3,3,3,3,3],
		big,
		]:
	b = qsort_naive(a)
	assert b == sorted(a)

	a2 = a[:]
	qsort(a2, 0, len(a2) - 1)
	assert a2 == b

print 'ok!'

File: algo/sorting/qsort.cpp
#include <stdlib.h>
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

int partition(vector<int>& v, int left, int right, int p_idx) {
	int p_val = v[p_idx];
	swap(v[p_idx], v[right]);
	int s_idx = left;
	for (int i = left; i < right; ++i) {
		if (v[i] <= p_val) {
			swap(v[i], v[s_idx]);
			s_idx += 1;
		}
	}
	swap(v[right], v[s_idx]);
	return s_idx;
}

void qsort(vector<int>& v, int left, int right) {
	if (right > left) {
		int p_idx = left + (right - left) / 2;
		int p_idx_new = partition(v, left, right, p_idx);
		qsort(v, left, p_idx_new - 1);
		qsort(v, p_idx_new + 1, right);
	}
}

void print(const vector<int>& v) {
	for (vector<int>::const_iterator i = v.begin(); i != v.end(); ++i)
		cout << *i << endl;
}

void fill(vector<int>& v) {
	for (vector<int>::iterator i = v.begin(); i != v.end(); ++i)
		*i = rand();
}

int main (int argc, char** argv) {
	const int n = 10;
	vector<int> v(n);
	fill(v);
	cout << endl << "unsorted" << endl;
	print(v);
	qsort(v, 0, n-1);
	cout << endl << "sorted" << endl;
	print(v);
	return 0;
}
